Add native callTracer and refactor based on new go-ethereum paradigms

ziogaschr opened the pull request:
This PR achieves the following:
[x] Refactor of parity like tracer added at core-geth, so as it follows the go-ethereum paradigms.
[x] Add new **native** `callParityTracer` that is much faster as it runs in Go.
[x] Add more tests for tracers.
[x] Update the compatibility layer with OE based on their latest changes, and running a full comparison test based on this [OE branch](https://github.com/ziogaschr/openethereum/commits/etc) which used OE v3.3.4.

--

`stateDiff` tracer will be simplified and be converted in a native tracer in a next PR.

ziogaschr committed:
eth/tracers/native: add BlockNumber in call_parity native tracer

ziogaschr committed:
eth/tracers/native: pass tracers.Context to Native tracers

ziogaschr committed:
eth/tracers/internal/tracetest/testdata/call_tracer_parity: eth/tracers/js/internal/tracers: rename tracer; fix precompiled check to happen on CaptureExit

ziogaschr committed:
eth/tracers/js/internal/tracers: add back old parity tracer, tagged as legacy (is slower)

ziogaschr committed:
eth/tracers/internal/tracetest/testdata/call_tracer_parity: remove `creationMethod` from tests

ziogaschr committed:
Revert "eth/tracers/internal/tracetest/testdata/call_tracer_parity: remove `creationMethod` from tests"

This reverts commit 0385ae42941d3e2cf6f1fdde339727c41e5f583f.

ziogaschr committed:
eth/tracers/js: bring back `getReturnData` from JS tracers (legacy)

ziogaschr committed:
eth/tracers/js/internal/tracers: fix precompiled check to happen on CaptureExit

ziogaschr committed:
eth/tracers/js/internal/tracers: 🩹 comment

ziogaschr committed:
eth/tracers/internal/tracetest: better reporting on failing tests; test all available Parity tracers

ziogaschr committed:
eth/tracers/native: rename `callParityTracer` to `callTracerParity` for native tracer

ziogaschr committed:
eth/tracers/native: eth/tracers/native: add missing transaction context to `callTracerParity` native tracer

ziogaschr committed:
eth/tracers/internal/tracetest/testdata/call_tracer_parity,eth/tracers/native: on parity tracer (Delegate|Static)call uses the value from parent, if zero

ziogaschr committed:
eth/tracers/internal/tracetest,eth/tracers/native: share `CallTrace*` struct definitions with tests

ziogaschr committed:
eth/tracers/internal/tracetest/testdata/call_tracer_parity: add `suicide` test for call_tracer_parity

ziogaschr committed:
eth/tracers/native: inline, omitting Parity tracer output fields on native tracer, with OE

ziogaschr committed:
eth/tracers/native: fix precompiled check to happen on CaptureExit on native Parity tracer

ziogaschr committed:
eth/tracers/internal/tracetest,eth/tracers/internal/tracetest/testdata/state_diff: use stateOverrides for stateDiff tracer (draft)

ziogaschr committed:
eth/tracers/js/internal/tracers: commit generated assets.go

ziogaschr committed:
eth/tracers/native: use the value from parent, if zero, only for Delegatecall

(to be cross-checked)

ziogaschr committed:
eth/tracers/internal/tracetest,eth/tracers/internal/tracetest/testdata/state_diff: fix/change stateDiff tracer tests to use stateOverrides

ziogaschr committed:
eth/tracers/native: 🩹

ziogaschr committed:
eth/tracers/internal/tracetest: eth/tracers/internal/tracetest: revert stateDiff test runner to use non Legacy tracer

ziogaschr committed:
eth/tracers: eth/tracers: remove vmctx.(Can)Transfer override for trace_call

go-ethereum instead added stateOverrides support

ziogaschr committed:
eth/tracers/internal/tracetest,eth/tracers/internal/tracetest/testdata/call_tracer_parity,eth/tracers/js/internal/tracers,eth/tracers/native: eth/tracers/internal/tracetest,eth/tracers/internal/tracetest/testdata/call_tracer_parity,eth/tracers/js/internal/tracers,eth/tracers/native: (omit)empty Parity call tracer fields based on OE

ziogaschr committed:
eth/tracers/internal/tracetest,eth/tracers/js/internal/tracers: eth/tracers/internal/tracetest,eth/tracers/js/internal/tracers: remove callTracerParityJs as we a same native tracer

and it will add tech debt for later in order to remove it

meowsbits commented on eth/tracers/internal/tracetest/testdata/call_tracer_parity/deep_calls.json:
```diff
@@ -120,7 +120,8 @@
       "result": {
-        "gasUsed": "0x12bb3"
+        "gasUsed": "0x12bb3",
+        "output": "0x"
```
`0x`'s are in! Do docs need updating?

ziogaschr replied:
No we don't. They were in already under curtain conditions, while now it's the default, except some conditions.

meowsbits committed:
eth/tracers/native: method modifies the pointer arg, should say so

Renames method to make it clear that the method
will modify the pointer argument.

Date: 2022-04-18 06:07:11-07:00
Signed-off-by: meows <b5c6@protonmail.com>

meowsbits commented on eth/tracers/internal/tracetest/testdata/call_tracer_parity/suicide.json:
```diff
@@ -0,0 +1,117 @@
+{
```
How did you make this test data? Should we consider including any test creation tools in the repo?

ziogaschr replied:
By using [this example code](https://github.com/etclabscore/core-geth/blob/d7543b9c21631466226e0f931edca13f92f9acca/eth/tracers/internal/tracetest/calltrace_test.go#L49-L96), but with the tracer name I need the test for.

Do you think we need to add anything more?

meowsbits replied:
Ah, looks great. Sorry I didn't see that earlier.

ziogaschr replied:
I even missed it for long time mate :)
